internal/backoff/backoff.go

```diff
@@ -30,47 +30,128 @@ import (
 
 // Strategy defines the methodology for backing off after a grpc connection
 // failure.
-//
 type Strategy interface {
 	// Backoff returns the amount of time to wait before the next retry given
 	// the number of consecutive failures.
 	Backoff(retries int) time.Duration
 }
 
-const (
-	// baseDelay is the amount of time to wait before retrying after the first
+// StrategyBuilder helps build an implementation of the Strategy interface,
+// with setters provided to set individual parameters of the backoff strategy.
+type StrategyBuilder interface {
+	// BaseDelay sets the amount of time to backoff after the first connection
 	// failure.
-	baseDelay = 1.0 * time.Second
-	// factor is applied to the backoff after each retry.
-	factor = 1.6
-	// jitter provides a range to randomize backoff delays.
-	jitter = 0.2
+	BaseDelay(time.Duration) StrategyBuilder
+	// Multiplier sets the the factor with which to multiply backoffs after a
+	// failed retry.
+	Multiplier(float64) StrategyBuilder
+	// Jitter sets the factor with which backoffs are randomized.
+	Jitter(float64) StrategyBuilder
+	// MaxDelay sets the upper bound of backoff delay.
+	MaxDelay(time.Duration) StrategyBuilder
+	// Build builds an implementation of the Strategy interface configured using
+	// the above setter methods.
+	Build() Strategy
+}
+
+// Default values for the different backoff parameters as defined in
+// https://github.com/grpc/grpc/blob/master/doc/connection-backoff.md.
+const (
+	baseDelay  = 1.0 * time.Second
+	multiplier = 1.6
+	jitter     = 0.2
+	maxDelay   = 120 * time.Second
 )
 
+// NewExponentialBuilder returns an implementation of the StrategyBuilder
+// interface used to build an exponential backoff strategy implementation, as
+// defined in
+// https://github.com/grpc/grpc/blob/master/doc/connection-backoff.md.
+func NewExponentialBuilder() StrategyBuilder {
```

Comment:
Go doesn't typically have builders or getters/setters.  Just structs with fields.

Why not simply export the fields in Exponential?

Defaults could be handled by something like:

```go
package backoff

var DefaultExponential = Exponential{<defaults>}
...
package grpc

s := DefaultExponential
s.BaseDelay = 10*time.Second
```

*or*

Make all the fields pointers and handle nil pointers in the `Backoff` function.  This is a bit more painful to use and handle, though you could make it easy to use with helpers like `func DurationPtr(d time.Duration) *time.Duration { return &d }`.